### PR TITLE
[CODE] Refactoring base.py sort model to be cleaner and easier to read. 

### PR DESCRIPTION
**Pull request by ughstudios**
- Some of the code in base.py is very bad and will probably need to be changed again. However, I added type hints and removed some bad code and cleaned up a bit of the logic.

#1278 

I wrote a comment on this last pull request. 

## Checklist

- [x] Keep pull requests small so they can be easily reviewed.
- [ ] Update the documentation, please follow the [numpy style guide](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] Update tests.
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `feature`, `ui`, `change`, `documentation`, `breaking`, `ci`
      as they show up in the changelog.
- [ ] Link this PR to related issues by using [closing keywords](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).
- [x] Add a milestone to the PR (and related issues, if any) for the intended release.
- [x] Request a review from another developer.

### CODE REVIEW

**Comment by cmutel** [comment]
This is a good first step but the logic here is still broken (not your fault). The earlier call replaces only `NaN` with an empty string. Assuming that missing values are zero is maybe OK, but WRT sorting I would put missing values (i.e. ones which don't have an explicit zero) below everything, including below negative numbers.

The basic problem is that earlier in the code there is an explicit inclusion of string types into a numeric array. Once this happens there is not reliable sorting. Or, in other words, this whole function should not need to exist.

**Comment by marc-vdm** [comment]
> The basic problem is that earlier in the code there is an explicit inclusion of string types into a numeric array.

Agree, plan is to update how results are processed after 2.10 sometime anyway, so this was not a priority. 

There's a related issue here: #887

File: activity_browser/ui/tables/models/base.py [hunk 1]
```diff
@@ -420,18 +420,24 @@ class ABSortProxyModel(QSortFilterProxyModel):
...
     """
-    def lessThan(self, left, right) -> bool:
+    def lessThan(self, left: QModelIndex, right: QModelIndex) -> bool:
         """Override to sort actual data, expects `left` and `right` are comparable.
 
         If `left` and `right` are not the same type, we check if numerical and empty string are compared, if that is the
         case, we assume empty string == 0.
         Added this case for: https://github.com/LCA-ActivityBrowser/activity-browser/issues/1215"""
-        left_data = self.sourceModel().data(left, 'sorting')
-        right_data = self.sourceModel().data(right, 'sorting')
```
Comment (marc-vdm) [inline activity_browser/ui/tables/models/base.py:430]: If you don't keep the `sorting` argument, the sort is done based on what is in the view, not the actual data (which is the reason `QSortFilterProxyModel` was subclassed in the first place). 

e.g. timestamps in a human readable format are not properly sorted if this argument is removed (#1151)


please restore.

File: activity_browser/ui/tables/models/base.py [hunk 2]
```diff
@@ -420,18 +420,24 @@ class ABSortProxyModel(QSortFilterProxyModel):
...
+        left_data = self.sourceModel().data(left)
+        right_data = self.sourceModel().data(right)
+        if not left_data and not right_data:
```
Comment (marc-vdm) [inline activity_browser/ui/tables/models/base.py:431]: Please make all follow-up checks `elif`, that makes it clearer this is one test case, not multiple in a row.

File: activity_browser/ui/tables/models/base.py [hunk 3]
```diff
@@ -420,18 +420,24 @@ class ABSortProxyModel(QSortFilterProxyModel):
...
-        elif type(left_data) in (int, float) and type(right_data) is str and right_data == "":
+
+        if type(left_data) in (int, float) and not right_data:
             return left_data < 0
-        elif type(right_data) in (int, float) and type(left_data) is str and left_data == "":
+
+        if type(right_data) in (int, float) and not left_data:
             return 0 < right_data
+
         raise ValueError(f"Cannot compare {left_data} and {right_data}, incompatible types.")
```
Comment (marc-vdm) [inline activity_browser/ui/tables/models/base.py:443]: I found a missing test case that wasn't in the original code, thanks!

Please also add testing between `str` and `None` in addition to numeric and `None`. Example in UI: 2 exchanges in an activity, one with a comment and one without.

**Review by marc-vdm — CHANGES_REQUESTED** [review]
Thanks for the PR! See separate comments for requested changes. If you have questions/comments, let me know.

In addition to requested code-changes, please update the name of your PR to something more easily understandable so we can understand/retrace from merge history/git blame what's happened and why, see the [pull-requests section of CONTRIBUTING.MD](https://github.com/LCA-ActivityBrowser/activity-browser/blob/main/CONTRIBUTING.md#pull-requests) for some pointers.

Finally, please don't remove the PR checklist when opening PRs, the checklist makes it easier for us to make sure the administrative side is roughly in order.

Thanks!

File: activity_browser/ui/tables/models/base.py (hunk 1 above)
Comment (marc-vdm) [reply activity_browser/ui/tables/models/base.py:430]: > It's not normally good practice to be passing strings like this

Ok, what _is_ good practice in this case? Happy to hear suggestions.

**Comment by marc-vdm** [comment]
>I'll also look into seeing where we could make a "real" fix to the issue.

I'd suggest keeping this low on your priorities, we have on the planning to make major changes to results.

File: activity_browser/ui/tables/models/base.py (hunk 1 above)
Comment (marc-vdm) [reply activity_browser/ui/tables/models/base.py:430]: The `'sorting'` string is passed as `role` argument, Qt just doesn't provide a role for what we are doing, so we created a custom one like this. I could have written a custom role, but just passing a string is less of a mess IMO

**Comment by marc-vdm** [comment]
@ughstudios We are looking to make a `2.10` release soon-ish. Will you finish the remaining task or should we close this?
